keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have terminated all running sidecar containers will also be terminated with a reason of `Completed`.
+
+For example, in a job you can have as many sidecars as you like and all the running sidecars will be terminated when all the non-sidecar containers have terminated, causing your job to finish.
+If all the non-sidecars `Completed` the pod status will be `Succeeded`, if any of the non-sidecars terminated with `Error` the pod status will be `Failed` and if any of the sidecars terminated with `Error` before the non-sidecars terminated, the pod status will be `Failed`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+//TODO
+
+### Risks and Mitigations
+
+You could set all containers to be `sidecar: true`, this seems wrong, so maybe the api should do a validation check that at least one container is not a sidecar.
+Init containers would be able to have `sidecar: true` applied to them as it's an additional field to the container spec, there's no situation where this would make sense as init containers are executed sequentially. We could get around this by having the api throw a validation error if you try to use this field on an init container.
+
+## Graduation Criteria
+
+//TODO
+
+## Implementation History
+
+//TODO
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+
+## Alternatives
+
+One alternative would be to have a new field in the Pod Spec of `sidecarContainers:` where you could define a list of sidecar containers, however this would require a lot more work in terms of updating tooling to support this.
+Another alternative would be to change the Job Spec to have a `primaryContainer` field to tell it which containers are important. However I feel this is perhaps too specific to job when this Sidecar concept could be useful in other scenarios.
```

Comment:
Sidecards are very useful outside of the `(Cron)Job` context so I agree that this should have a larger scope for all `Pod`s.

Reply:
Another advantage of a `sidecarContainers` part of the pod spec might be to make them useable with `initContainers`. 

Many of the typical types of sidecars are for logging, monitoring or proxies, which can be useful to init containers as well. Currently its not possible to run a sidecar with them at all, even shell magic doesnt help as they are run one at a time.

I wrote in https://github.com/kubernetes/kubernetes/issues/25908#issuecomment-404282114 of example of migrations running in initContainer and needing access to sql proxy sidecar.

Reply:
I think the generality of this proposal is not related to which polarity the flag is.  We should figure out which direction makes the most sense to users.  In my mind "is sidecar" makes the most sense because the default would be false and  injected sidecars know they are injected, and only need to say something about themseleves.